DeviceAdministration/Infrastructure.UnitTests/Infrastructure/DeviceRegistryRepositoryTests.cs

```diff
@@ -0,0 +1,252 @@
+        }
+
+        [Fact]
+        public async Task GetDeviceAsync_should_get_device_from_client()
+        {
+            var devices = this._fixture.Create<List<DeviceModel>>();
+            var deviceId = devices.First().DeviceProperties.DeviceID;
+
+            this._mockDocumentDBClient
+                .Setup(x => x.QueryAsync())
+                .ReturnsAsync(devices.AsQueryable());
+
+            var device = await this._deviceRegistryRepository.GetDeviceAsync(deviceId);
+
+            Assert.NotNull(device);
+            Assert.Same(devices[0], device);
+        }
+
+        [Fact]
+        public async Task GetDeviceAsync_should_return_null_when_device_id_is_not_found()
+        {
+            var devices = this._fixture.Create<List<DeviceModel>>();
+
+            this._mockDocumentDBClient
+                .Setup(x => x.QueryAsync())
+                .ReturnsAsync(devices.AsQueryable());
+
+            var device = await this._deviceRegistryRepository.GetDeviceAsync("foobarbaz");
+
+            Assert.Null(device);
+        }
+
+        [Fact]
+        public async Task AddDeviceAsync_should_throw_argument_null_when_device_is_null()
+        {
+            await Assert.ThrowsAsync<ArgumentNullException>(() => this._deviceRegistryRepository.AddDeviceAsync(null));
+        }
+
+        [Fact]
+        public async Task AddDeviceAsync_should_throw_device_exists_exception_when_device_with_same_id_exists()
+        {
+            var devices = this._fixture.Create<List<DeviceModel>>();
+            this._mockDocumentDBClient
+                .Setup(x => x.QueryAsync())
+                .ReturnsAsync(devices.AsQueryable());
+
+            await Assert.ThrowsAsync<DeviceAlreadyRegisteredException>(() => this._deviceRegistryRepository.AddDeviceAsync(devices.First()));
+        }
+
+        [Fact]
+        public async Task AddDeviceAsync_should_add_device_to_db()
+        {
+            var device = this._fixture.Create<DeviceModel>();
+            await this._deviceRegistryRepository.AddDeviceAsync(device);
+            this._mockDocumentDBClient.Verify(x => x.SaveAsync(device));
+        }
+
+        [Theory]
+        [InlineData(null)]
+        [InlineData("")]
+        public async Task RemoveDeviceAsync_should_throw_argument_null_when_device_id_is_not_valid(string deviceId)
+        {
+            await Assert.ThrowsAsync<ArgumentNullException>(() => this._deviceRegistryRepository.RemoveDeviceAsync(deviceId));
+        }
+
+        [Fact]
+        public async Task RemoveDeviceAsync_should_throw_device_not_found_when_device_does_not_exist()
+        {
+            this._mockDocumentDBClient
+                .Setup(x => x.QueryAsync())
+                .ReturnsAsync(Enumerable.Empty<DeviceModel>().AsQueryable());
+
+            await Assert.ThrowsAsync<DeviceNotRegisteredException>(() => this._deviceRegistryRepository.RemoveDeviceAsync("foobar"));
+
+            this._mockDocumentDBClient.Verify();
```

Comment:
Verify auto verifies any previously .Setup calls. So no explicit Verify(x = x.QueryAsync()) is required.